Object.prototype should be an immutable prototype exotic object.

dilijev opened the pull request:
This means that any attempts to set the [[Prototype]] slot of the Object Prototype Object (the object which is initially pointed to by `Object.prototype`) notably by using `Object.prototype.__proto__` or `Object.setPrototypeOf(Object.prototype, otherObject)` should fail with a `TypeError` as per spec.

This is the only exotic behavior of this object.

I chose to implement this by adding the virtual method `IsProtoImmutable` to `RecyclableObject` with the default return value `false`, and to override this method in `ObjectPrototypeObject` to return `true`. This allows for a future implementation of `Object.setImmutablePrototype` which would allow other objects to exhibit this behavior. A flag could be added which can be set, and this method would return the value of that flag. In that case, the implementation in `ObjectPrototypeObject` should remain hardcoded to false to avoid accidentally allowing the value if `IsProtoImmutable` to change.

See [ES2016 19.1.3](https://tc39.github.io/ecma262/#sec-properties-of-the-object-prototype-object):
The Object prototype object is the intrinsic object %ObjectPrototype%.
The Object prototype object is an immutable prototype exotic object.
See [ES2016 9.4.7](https://tc39.github.io/ecma262/#sec-immutable-prototype-exotic-objects):
An immutable prototype exotic object is an exotic object that has an immutable [[Prototype]] internal slot.

Fixes #261

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/microsoft/chakracore/878)
<!-- Reviewable:end -->

dilijev commented:
@bterlson @ljharb @abchatra @curtisman 
Please review?

goyakin commented on lib/Parser/rterrors.h:
```diff
@@ -275,6 +275,8 @@ RT_ERROR_MSG(JSERR_DeletePropertyWithSuper, 5146, "Unable to delete property '%s
 RT_ERROR_MSG(JSERR_DetachedTypedArray, 5147, "%s: The ArrayBuffer is detached.", "The ArrayBuffer is detached.", kjstTypeError, 0)
 RT_ERROR_MSG(JSERR_AsmJsCompileError, 5148, "%s: Compiling asm.js failed.", "Compiling asm.js failed.", kjstError, 0)
 
+RT_ERROR_MSG(JSERR_ImmutablePrototypeSlot, 5149, "%s: Can't set the prototype slot of this object.", "Can't set the prototype slot of this object.", kjstTypeError, 0)
```
I wouldn't mention _slot_ here since users don't need to know the language internals. I think it would be better to say `Can't set the prototype of this object` instead.

dilijev replied:
Makes sense. I'll make this change.

goyakin commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -176,6 +176,16 @@ namespace Js
 
+        if (object->IsProtoImmutable())
+        {
+            // 19.1.3:
```
Section numbers change between spec versions, so you should specify "ES2016" here.

dilijev replied:
Fixing this as well.

akroshg commented on test/es7/immutable-prototype.js:
```diff
@@ -0,0 +1,77 @@
+            assert.areEqual(Object.prototype, Object.getPrototypeOf(a), "Object.getPrototypeOf(a) === Object.prototype")
+
+            assert.doesNotThrow(function () { eval("Object.setPrototypeOf(a, b)") },
+                "It should be okay to set the [[Prototype]] of an arbitrary object using __proto__.")
```
using *setPrototypeOf * ?

dilijev replied:
Thanks!

goyakin commented on test/es7/immutable-prototype.js:
```diff
@@ -0,0 +1,77 @@
+
+var tests = [
+    {
+        name: "Okay to set arbitrary object's [[Prototype]] using __proto__",
```
Looks like we already test this in `test/es6/proto_basic.js`. Look for the test with name `Change object.__proto__ to an object`.

dilijev replied:
Ah, I figured there would be a test already. Since I wanted to make sure the new behavior didn't break the old behavior I added these tests here, but I see that they are redundant so I'll remove them.

goyakin commented on test/es7/immutable-prototype.js:
```diff
@@ -0,0 +1,77 @@
+        }
+    },
+    {
+        name: "Okay to set arbitrary object's [[Prototype]] using Object.setPrototypeOf",
```
Already tested in `test/es6/ES6Proto.js`. Look for the test with name `Simple validation of Object.setPrototypeOf`.

dilijev replied:
Removing.

goyakin commented on test/es7/immutable-prototype.js:
```diff
@@ -0,0 +1,77 @@
+            var objectPrototypeObject = Object.getPrototypeOf(Object.prototype)
+            var b = Object.create(null)
+
+            assert.throws(function () { eval("Object.prototype.__proto__ = b") },
```
Is there a particular reason why `eval` used here instead of doing a plain assignment to `__proto__`?

dilijev replied:
I had only seen assert.throws that used eval. I thought it might have been required for some reason. I don't need it, I'll get rid of it.

goyakin commented on test/es7/immutable-prototype.js:
```diff
@@ -0,0 +1,77 @@
+            var objectPrototypeObject = Object.getPrototypeOf(Object.prototype)
+            var b = Object.create(null)
+
+            assert.throws(function () { eval("Object.setPrototypeOf(Object.prototype, b)") },
```
Can we directly call `setPrototypeOf` without `eval`?

dilijev replied:
Yes, I'll make the change.

goyakin commented:
Other than minor inline comments, everything else LGTM.

> I chose to implement this by adding the virtual method IsProtoImmutable to RecyclableObject with the default return value false, and to override this method in ObjectPrototypeObject to return true.

This would be useful when we implement `[[SetPrototypeOf]]` for [module namespace objects](https://tc39.github.io/ecma262/#sec-module-namespace-exotic-objects-setprototypeof-v) :+1:

dilijev commented:
There is a failing test because of defunct behavior (directly contradicts that `Object.prototype.__proto__` should be immutable).

dilijev commented:
Removed defunct test and updated spec references to specifically refer to ES2016. Thanks @goyakin.

/cc @jianchun @bterlson on test removal.

`Cyclic __proto__ ` was tested in another test in proto_basic.js which didn't involve Object.prototype. All of the other assignments don't make sense anymore since the [[Prototype]] slot is immutable so removing that unit test made sense.
